pulsar-common/src/main/java/org/apache/pulsar/common/protocol/PulsarHandler.java

```diff
@@ -152,5 +152,19 @@ public void cancelKeepAliveTask() {
      */
     protected abstract boolean isHandshakeCompleted();
 
+    /**
+     * Demo: [id: 0x2561bcd1, L:/10.0.136.103:58038 ! R:/240.240.0.5:6650].
+     * L: local Address.
+     * R: remote address.
+     */
+    @Override
+    public String toString() {
+        if (ctx == null) {
+            return "[ctx: null]";
+        } else {
+            return ctx.channel().toString();
```

Comment:
Is channel always non-null?

Reply:
Yes, in normal the `ctx` is typed `AbstractChannelHandlerContext`, and the variable `pipeline` with modifier `final` will be initialized in the constructor, and the variable `channel` of the pipeline is also initialized in its constructor.